Housekeeping module

Vysares opened the pull request:
Housekeeping module complete. Housekeeping data at any given time is store in latestHksample, which is a struct containing all the temperature and power data. Previous samples are stored in hkData, which is just an array of many structs. 
Also wrote a MATLAB script for generating the thermistor voltage/temperature lookup table. I'm not sure what realistic temperature range is, the the range of that lookup table and the values of all the min/max temperatures in config.hpp are just guesses for now.

Vysares committed:
houskeeping initial commit

Vysares committed:
update hkUtil hpp & cpp

Vysares committed:
Merge branch 'main' into housekeeping_module_dev

Vysares committed:
Update comUtil.cpp

Changed commandQueue to static type

Vysares committed:
Update Config.hpp

Added Housekeeping variables. Renamed SPI_MAX_SPEED to ADC_MAX_SPEED, since the max speed will be different for the flash.

Vysares committed:
update hkUtil and config

Vysares committed:
Update config.hpp

added housekeeping constants and teensy ADC constans for use with analogRead()

Vysares committed:
Merge branch 'main' into housekeeping_module_dev

Vysares committed:
Update config.hpp

renamed min and max safe temp variables. added expected PG signal range

Vysares committed:
housekeeping util functionality

Housekeeping module initial functionality complete

Vysares committed:
Squashed bugs

Vysares committed:
Create ThermistorTempCalc.m

added new MATLAB folder, and script to generate thermistor voltage/temp lookup table

jaretanderson commented on FSW/src/headers/config.hpp:
```diff
@@ -74,13 +85,40 @@ static TimedEvent sweepTimeoutEvent(SWEEP_TIMEOUT_MSEC);
 //      look at checkSweepChange() in timing.cpp for more info
 static Event sweepDirectionChangeEvent; 
 
+
 /* - - - - - - Command Handling Module - - - - - - */
 const int COMMAND_QUEUE_SIZE = 100;     // maximum number of commands the command queue can store.
 const int SERIAL_TIMEOUT_MSEC = 50;     // milliseconds, maximum time to wait for serial input
 
+
 /* - - - - - - Fault Mitigation Module - - - - - - */
-const int WD_RESET_INTERVAL = 100;  // watchdog feeding interval, ms
-const int WD_PULSE_DUR = 10;        // watchdog reset signal duration, MICROSECONDS!!!
+const int WD_RESET_INTERVAL_MSEC = 100;     // milliseconds, watchdog feeding interval
+const int WD_PULSE_DUR_MICROSEC = 10;       // microseconds, watchdog reset signal duration
```
I like all of the standardization you did in `config.hpp`, makes it much clearer what the constants mean

jaretanderson commented on FSW/src/headers/hkUtil.hpp:
```diff
@@ -0,0 +1,52 @@
+
+struct TempVoltagePair { float temperature, voltage; };
+
+// thermistor temperature/voltage lookup table
```
```suggestion
// thermistor temperature (deg C) / voltage (Volts) lookup table
```

jaretanderson commented on FSW/src/util/hkUtil.cpp:
```diff
@@ -0,0 +1,184 @@
+ * Outputs:
+ *  None
+ */
+void setHeater()
+{
+    if (latestHkSample.opticsTemp >= HEATER_TEMP_HIGH)
+    {
+        digitalWrite(PIN_HEAT, LOW); // turn heater off
+    }
+    else if (latestHkSample.opticsTemp <= HEATER_TEMP_LOW)
+    {
+        digitalWrite(PIN_HEAT, HIGH); // turn heater on
+    }
+}
```
do you think it is necessary to implement a thermostat algorithm instead of a direct if-statement? 

Right now if the temp is hovering right around `HEATER_TEMP_LOW`/`HIGH` then we could end up turning the heater on and off in a fast oscillating manner every time the loop is called, which could potentially cause electrical noise

Depending on the frequency that `handleHousekeeping()` is called with though (i,e. if it is ~1Hz) then this probably isn't something we need to worry about.

Vysares replied:
If the temp drops below HEATER_TEMP_LOW the heater is turned on and it will stay on until the temperature increases above HEATER_TEMP_HIGH, then it will turn off and stay off until the temp drops back down. That way we get a slow oscillation between the LOW and HIGH temps. As long as the range between them is not tiny this shouldn't trigger too rapidly.

JashanChopra replied:
I don't really see a scenario where the range is exactly on the low or high and it triggers a bunch, but it would probably not be a bad thing to trigger a warning if it hits a bunch, I'm not really sure.

jaretanderson reviewed (approved):
Donavon was doing some research on thermal maxes/mins for the thermal test plan, so we should talk to him and get confirmation on the config numbers but the logic here looks good to me

Vysares committed:
Update FSW/src/headers/hkUtil.hpp

specified units in lookup table description

Co-authored-by: Jaret Anderson <49298404+jaretanderson@users.noreply.github.com>

Vysares committed:
Merge branch 'main' into housekeeping_module_dev

Vysares committed:
Update config.hpp

Fixed merge Shenanigans

Vysares committed:
Update hkUtil.hpp

Updated comments

Vysares committed:
Merge branch 'housekeeping_module_dev' of https://github.com/Vysares/NanoSAM2 into housekeeping_module_dev

JashanChopra commented:
> Donavon was doing some research on thermal maxes/mins for the thermal test plan, so we should talk to him and get confirmation on the config numbers but the logic here looks good to me

Donavon said he found -25C to 70C. For this semester, they won't test outside of that range. For actual FSW, you'd probably want to go below that range so that you can check for when the temperature gets too low, and if it isn't corrected quickly, perhaps a transition to safe mode can be applied.

JashanChopra commented on FSW/src/headers/config.hpp:
```diff
@@ -69,13 +80,45 @@ static Event saveBufferEvent;
+
+
+/* - - - - - - Housekeeping Module - - - - - - */
+const int HK_SAMPLES_TO_KEEP = 100;   // number of previous housekeeping samples to keep
```
How much time would this cover? I didn't do the math. I imagine we should keep _all_ the housekeeping data between downlink periods for debugging. Data science might want to link temperatures with the data on the ground.

Vysares replied:
Very good point. It should be somewhere around ~5000 elements large depending on the orbit. With six floats per sample all the housekeeping data will be about 0.5Mbits

JashanChopra commented on FSW/src/headers/config.hpp:
```diff
@@ -69,13 +80,45 @@ static Event saveBufferEvent;
+const int HK_SAMPLES_TO_KEEP = 100;   // number of previous housekeeping samples to keep
+
+// heater cutoff temperatures
+const float HEATER_TEMP_LOW = 0;     // celsius, heater will turn on at or below this temp
```
I think Danny was saying something about the fact that the heater should basically always be on. Probably still good to have the upper limit though. 

I would design the heater to be always on by default, but I don't really like the idea of that for testing, draws a lot of power.

Vysares replied:
I added a bool `forceHeaterOn` that `setHeater()` will check first, and if true will just force the heater on all the time.

JashanChopra commented on FSW/src/headers/config.hpp:
```diff
@@ -69,13 +80,45 @@ static Event saveBufferEvent;
+// safe temperature range
+const float OPTICS_TEMP_MIN_SAFE = -20; // celsius, minimum safe photodiode temp
+const float OPTICS_TEMP_MAX_SAFE = 50;  // celsius, maximum safe photodiode temp
+const float BOARD_TEMP_MIN_SAFE = 0;    // celsius, minimum safe board temp
```
-20

JashanChopra commented on FSW/src/headers/config.hpp:
```diff
@@ -69,13 +80,45 @@ static Event saveBufferEvent;
+const float OPTICS_TEMP_MIN_SAFE = -20; // celsius, minimum safe photodiode temp
+const float OPTICS_TEMP_MAX_SAFE = 50;  // celsius, maximum safe photodiode temp
+const float BOARD_TEMP_MIN_SAFE = 0;    // celsius, minimum safe board temp
+const float BOARD_TEMP_MAX_SAFE = 50;   // celsius, maximum safe board temp
```
70

JashanChopra commented on FSW/src/headers/hkUtil.hpp:
```diff
@@ -0,0 +1,52 @@
+float voltageToTemp(float voltage);
+void timeSortHkData();
+
+struct HousekeepingData // stores one point of housekeeping data
```
Make sure structs like these are documented well outside of the code for reference when future teams (or us later in testing) are looking at the data.

Vysares replied:
I will update the readme with info on all the data structures, and more generally to provide an overview of how the software fits together for people who are not me and Jaret.

Vysares replied:
See new issue

JashanChopra commented on FSW/src/util/hkUtil.cpp:
```diff
@@ -0,0 +1,184 @@
+
+
+/* Module Variable Definitions */
+static HousekeepingData hkData[HK_SAMPLES_TO_KEEP] = {}; // array to store housekeeping data
```
`housekeepingData` for camelCase consistency.

Vysares replied:
changed.

JashanChopra commented on FSW/src/util/hkUtil.cpp:
```diff
@@ -0,0 +1,184 @@
+
+    // Check if values are in acceptable ranges
+    // TODO: Log a fault for every case
+    if (latestHkSample.opticsTemp > OPTICS_TEMP_MAX_SAFE) { } // optics temp too high
```
Seems like there's a redundancy between these first two if-else-if statements and the checks in `setHeater()`. I'd either just turn the heater on and off inside `sampleHousekeepingData` or I would put the fault logging in `setHeater()`.

Vysares replied:
There is a difference between the MAX/MIN safe temperatures and the LOW/HIGH heater trigger temperatures. like this:

|MIN_SAFE| <- - - - - |HEATER_LOW(turn on)| <- -(target range)- -> |HEATER_HIGH(turn off)| - - - - - -> |MAX_SAFE|
                                                                            
setHeater checks temperature against HEATER_TEMP_LOW and HEATER_TEMP_HIGH only, and we would expect the temperature to drift a little bit outside that range while the heater is changing state, but if for some reason the temperature drifts well beyond the target range and outside of the safe range we will want to log a fault.

JashanChopra commented on FSW/src/util/hkUtil.cpp:
```diff
@@ -0,0 +1,184 @@
+    if (latestHkSample.opticsTemp > OPTICS_TEMP_MAX_SAFE) { } // optics temp too high
+    else if (latestHkSample.opticsTemp < OPTICS_TEMP_MIN_SAFE) { } // optics temp too low
+
+    if (latestHkSample.analogTemp > BOARD_TEMP_MAX_SAFE) { } // analog board temp too high
```
If any of these are outside of the condition, a transition to safe mode may be validated.

Vysares replied:
If any of these trigger, a fault will be logged, and the fault will generate a command to enter safe mode.

JashanChopra commented on FSW/src/util/hkUtil.cpp:
```diff
@@ -0,0 +1,184 @@
+ * Outputs:
+ *  temperature - celsius, the temperature corresponding to thermistor voltage reading
+ */
+float voltageToTemp(float voltage)
```
The table you created, using the MATLAB script (which looks good) is only valid for the analog and digital thermistors. As per these equations: 
![image](https://user-images.githubusercontent.com/25249287/108008020-93d9ee80-6fbc-11eb-9188-9feb8b47f5e1.png)

Recall that the optics thermistor uses a different set of equations: which is recreated via this transfer table: 
![image](https://user-images.githubusercontent.com/25249287/108008065-ad7b3600-6fbc-11eb-8472-d621d58a9a4b.png)

Vysares replied:
Thanks, seems like I missed this. Will implement.

JashanChopra commented on FSW/src/util/hkUtil.cpp:
```diff
@@ -0,0 +1,184 @@
+    
+    // find lower adjacent entry
+    int i = 0;
+    while (thermLookup[i].voltage > voltage && i < thermLookupSize - 1) { i++; }
```
This looks good. If these calculations do prove to be too intense at the required rate, you can just compare raw voltage values (just take the corresponding voltages at the triggering temperatures). Then we provide good documentation on how to convert voltages to temperatures on the ground for science.

JashanChopra commented on FSW/src/util/hkUtil.cpp:
```diff
@@ -0,0 +1,184 @@
+ * Outputs:
+ *  none
+ */
+void timeSortHkData()
```
How similar is this to the similar function that sorts the science data by ascending? Can they both be used for the same purpose, or does the `struct` nature of the HK data require this new function?

Vysares replied:
Very similar. I had the same thought, but C++ really likes to know exactly what kind of data it is working with, so making generalized timeSortArray function would be more complicated and would likely use more memory than just having the two functions. 
This works for now, but in the future I am thinking of maybe replacing each array with a std::vector. They use more memory, but they are dynamically scalable, and have all kinds of awesome functions like push_back() and erase(), which would allow us to keep the vector sorted all the time.

JashanChopra reviewed (changes requested):
Transfer table for the optics thermistor must be implemented. The rest looks solid.

Vysares committed:
Update safe temperature ranges

Vysares committed:
update hkUtil.cpp

Added sample time field to housekeeping data struct.
Implemented forceHeaterOn, which can be set in config, that will force the heater to be always on

Vysares committed:
Added optics temp conversion and new commands

-Added new function to find the optics thermistor temp from voltage.
-Added new config values for thermistor calibration and gain.
-Added several new commands.
-Updated therm lookup table to span a wider range of temperatures

Vysares committed:
Removed reference to timingDeclarations.hpp

JashanChopra reviewed (approved):
Nice work, thanks for the changes.

jaretanderson commented on FSW/src/headers/config.hpp:
```diff
@@ -69,13 +80,51 @@ static Event saveBufferEvent;
 static TimedEvent sunriseTimerEvent(WINDOW_LENGTH_MSEC);
 static TimedEvent sweepTimeoutEvent(SWEEP_TIMEOUT_MSEC);
 
+// FUTURE TEAMS: this event is invoked when the ADCS should switch its sweep direction
+//   so link your ADCS module with this event to tell it when to switch direction 
+//      look at checkSweepChange() in timing.cpp for more info
+static Event sweepDirectionChangeEvent; 
+
+
```
This event got moved into the `timingClass.hpp` header, it shouldn't be in this file anymore. I wrote a note to circle back and look at the file once you merge it into main, if this shows back up then we will have to remove it, but the merge may take care of that.

Vysares committed:
Squashed bugs
